File: clients/android/app/src/main/kotlin/com/lextures/android/core/lms/LmsFeatureModelsMarketplace.kt
package com.lextures.android.core.lms

import kotlinx.serialization.Serializable

/** Authenticated in-app course marketplace models (plan MKT6). */

@Serializable
data class MarketplaceCourse(
    val id: String,
    val slug: String,
    val courseCode: String,
    val title: String,
    val description: String = "",
    val heroImageUrl: String? = null,
    val category: String? = null,
    val level: String? = null,
    val language: String = "en",
    val priceCents: Int = 0,
    val priceCurrency: String = "usd",
    val listPriceCents: Int? = null,
    val enrollmentCount: Int = 0,
    val averageRating: Double? = null,
    val ratingCount: Int = 0,
    val instructorName: String? = null,
    val createdAt: String = "",
    val owned: Boolean = false,
)

@Serializable
data class MarketplaceSearchResponse(
    val courses: List<MarketplaceCourse> = emptyList(),
    val total: Int = 0,
    val nextCursor: String = "",
)

@Serializable
data class MarketplaceCategory(
    val category: String,
    val count: Int = 0,
)

@Serializable
data class MarketplaceCategoriesResponse(
    val categories: List<MarketplaceCategory> = emptyList(),
)

@Serializable
data class MarketplaceWhatsIncluded(
    val moduleCount: Int = 0,
    val itemCount: Int = 0,
    val estimatedDurationMinutes: Int? = null,
)

@Serializable
data class MarketplaceRating(
    val average: Double? = null,
    val count: Int = 0,
)

@Serializable
data class MarketplaceCourseDetail(
    val course: MarketplaceCourse,
    val owned: Boolean = false,
    val priceCents: Int = 0,
    val priceCurrency: String = "usd",
    val listPriceCents: Int? = null,
    val whatsIncluded: MarketplaceWhatsIncluded = MarketplaceWhatsIncluded(),
    val rating: MarketplaceRating = MarketplaceRating(),
)

@Serializable
data class MarketplaceClaimResult(
    val enrolled: Boolean = false,
    val entitlementId: String = "",
    val alreadyOwned: Boolean = false,
    val firstItemId: String? = null,
    val courseCode: String = "",
)

@Serializable
data class CourseCatalogListing(
    val isPublic: Boolean = false,
    val category: String? = null,
    val difficultyLevel: String? = null,
    val language: String = "en",
    val priceCents: Int = 0,
    val priceCurrency: String = "usd",
    val slug: String = "",
    val marketplaceListed: Boolean = false,
    val publishState: String = "draft",
    val activePurchaseCount: Int = 0,
)

@Serializable
data class CourseCatalogListingResponse(
    val listing: CourseCatalogListing,
)

@Serializable
data class CourseCatalogListingPutBody(
    val isPublic: Boolean = false,
    val category: String? = null,
    val difficultyLevel: String? = null,
    val language: String = "en",
    val priceCents: Int = 0,
    val priceCurrency: String = "usd",
    val slug: String = "",
    val marketplaceListed: Boolean? = null,
)

File: clients/ios/Lextures/Core/LMS/LMSFeatureModelsMarketplace.swift
import Foundation

// MARK: - Authenticated course marketplace (MKT6)

struct MarketplaceCourse: Codable, Identifiable, Hashable {
    var id: String
    var slug: String
    var courseCode: String
    var title: String
    var description: String
    var heroImageUrl: String?
    var category: String?
    var level: String?
    var language: String
    var priceCents: Int
    var priceCurrency: String
    var listPriceCents: Int?
    var enrollmentCount: Int
    var averageRating: Double?
    var ratingCount: Int
    var instructorName: String?
    var createdAt: String
    var owned: Bool
}

struct MarketplaceSearchResponse: Codable {
    var courses: [MarketplaceCourse]?
    var total: Int?
    var nextCursor: String?
}

struct MarketplaceCategory: Codable, Identifiable, Hashable {
    var category: String
    var count: Int

    var id: String { category }
}

struct MarketplaceCategoriesResponse: Codable {
    var categories: [MarketplaceCategory]?
}

struct MarketplaceWhatsIncluded: Codable, Hashable {
    var moduleCount: Int
    var itemCount: Int
    var estimatedDurationMinutes: Int?
}

struct MarketplaceRating: Codable, Hashable {
    var average: Double?
    var count: Int
}

struct MarketplaceCourseDetail: Codable {
    var course: MarketplaceCourse
    var owned: Bool
    var priceCents: Int
    var priceCurrency: String
    var listPriceCents: Int?
    var whatsIncluded: MarketplaceWhatsIncluded
    var rating: MarketplaceRating
}

struct MarketplaceClaimResult: Codable {
    var enrolled: Bool
    var entitlementId: String
    var alreadyOwned: Bool?
    var firstItemId: String?
    var courseCode: String
    /// True when enrollment came from a 100% coupon grant (MKTC.3 / MKTC.6).
    var grantedFree: Bool?
}

struct MarketplaceCheckoutResult: Codable {
    var sessionId: String?
    var checkoutUrl: String?
    var alreadyOwned: Bool?
    var courseCode: String?
    var courseId: String?
    var grantedFree: Bool?
    var chargedCents: Int?
    var currency: String?
}

// MARK: - Coupons (MKTC.6)

struct CouponPreview: Codable, Equatable {
    var applied: Bool
    var code: String
    var reason: String
    var listPriceCents: Int
    var discountCents: Int
    var chargedCents: Int
    var currency: String
    var freeAfterDiscount: Bool
    var endsAt: String?
    var seatsRemaining: Int?
    var clampedToFree: Bool?

    static func rejected(code: String, reason: String, listPriceCents: Int, currency: String) -> CouponPreview {
        CouponPreview(
            applied: false,
            code: code,
            reason: reason,
            listPriceCents: listPriceCents,
            discountCents: 0,
            chargedCents: listPriceCents,
            currency: currency,
            freeAfterDiscount: false
        )
    }
}

struct MarketplaceCouponPreviewBody: Codable {
    var code: String
}

struct MarketplaceCouponCodeBody: Codable {
    var couponCode: String?
}

struct CoursePurchase: Codable, Identifiable, Hashable {
    var courseCode: String
    var courseId: String
    var title: String
    var priceCents: Int
    var currency: String
    var source: String
    var acquiredAt: String
    var receiptUrl: String?
    var entitlementId: String

    var id: String { entitlementId }
}

struct CoursePurchasesResponse: Codable {
    var purchases: [CoursePurchase]?
}

struct CourseCatalogListing: Codable, Hashable {
    var isPublic: Bool
    var category: String?
    var difficultyLevel: String?
    var language: String
    var priceCents: Int
    var priceCurrency: String
    var slug: String
    var marketplaceListed: Bool
    var publishState: String
    var activePurchaseCount: Int
}

struct CourseCatalogListingResponse: Codable {
    var listing: CourseCatalogListing
}

struct CourseCatalogListingPutBody: Codable {
    var isPublic: Bool
    var category: String?
    var difficultyLevel: String?
    var language: String
    var priceCents: Int
    var priceCurrency: String
    var slug: String
    var marketplaceListed: Bool?
}
